jdaviz/configs/mosviz/plugins/parsers.py

```diff
@@ -463,90 +454,148 @@ def mos_image_parser(app, data_obj, data_labels=None, share_image=0,
-            gratings = [x[0].header.get("GRATING", FALLBACK_NAME) for x in data_obj]
+            _add_to_table(app, filters_gratings, "Filter/Grating")
 
-            filters_gratings = [(f+'/'+g) for f, g in zip(filters, gratings)] * repeat
+        # source name and coordinates are taken from image headers, if present
+        if "Images" in current_columns:
+            ra = query_metadata_by_component(app, "OBJ_RA", "Images", float("nan"))
+            dec = query_metadata_by_component(app, "OBJ_DEC", "Images", float("nan"))
+            _add_to_table(app, ra, "R.A.")
+            _add_to_table(app, dec, "Dec.")
 
-            [x.close() for x in data_obj]
 
-        # source name can be taken from 1d spectra
-        elif spectra and sp1d:
-            names = _get_source_identifiers_by_hdu([x[0] for x in data_obj], ids) * repeat
+def query_metadata_by_component(app, keys, data_type, fallback=FALLBACK_NAME):
+    """
+    Searches and returns metadata values for a specific data component (type)
 
-        # source name and coordinates are taken from image headers, if present
-        else:
-            ra = [x[0].header.get("OBJ_RA", float("nan")) for x in data_obj] * repeat
-            dec = [x[0].header.get("OBJ_DEC", float("nan")) for x in data_obj] * repeat
-            names = _get_source_identifiers_by_hdu([x[0] for x in data_obj], ids) * repeat
+    If multiple keys are specified, the first found will be returned based on the order
+    keys are provided
 
-        [x.close() for x in data_obj]
+    Parameters
+    ----------
+    app : `~jdaviz.app.Application`
+        The application-level object used to reference the viewers.
+    keys : str or list
+        Metadata keywords to search for
+    data_type : str
+        The type of data to search for the Source ID keywords in. Can be one of:
+        "1D Spectra", "2D Spectra", "Images"
+    fallback : str
+        The value to return in the event the keyword could not be found
+    """
 
-    else:
-        # TODO: Come up with more robust metadata parsing, perhaps from
-        # the spectra files.
-        warnings.warn("Could not parse metadata from input images.")
-        return
+    metadata_list = list()
+    # If the user only provided a key that's not already in a container or list, put it in
+    # one for the upcoming loop
+    if not isinstance(keys, Iterable) or isinstance(keys, str):
+        keys = [keys]
 
-    with app.data_collection.delay_link_manager_update():
+    for data in app._jdaviz_helper.get_column(data_type):
+        meta = app.data_collection[data].meta
 
-        # this conditional must mirror the one above
-        if spectra and not sp1d:
-            _add_to_table(app, filters_gratings, "Filter/Grating")
-        elif spectra and sp1d:
-            _add_to_table(app, names, "Identifier")
+        # Search all given keys to see if they exist. Return the first hit
```

Comment:
So is the use case for this "the metadata I want might be under one of multiple keys" rather than "give me the metadata for every key in this list"?

Reply:
I think that's correct? Definitely the first part is correct. The "Identifier" field is the easiest to trace the logic path through; we previously pulled the Identifier through either "SOURCEID" or in some cases "OBJECT". Rather than manually hardcoding that loop, you can just provide this method a list of `['SOURCEID', 'OBJECT']` and it will search those keys for a value. 

On top of that, the order you provide the list also specifies the priority order to return. In the above example, if both `SOURCEID` and `OBJECT` are present, then it will return `SOURCEID`

Reply:
Cool, just checking that my understanding was correct.